Test failure on 0.3.45: test_siblings_cant_talk_parent

### Description

In Debian's build of 0.3.45, we are getting this failure:

```
======================================================================
ERROR: test_siblings_cant_talk_parent (router_test.UnidirectionalTest.test_siblings_cant_talk_parent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/reproducible-path/python-mitogen-0.3.45/tests/router_test.py", line 438, in test_siblings
_cant_talk_parent
    parent.call(test_siblings_cant_talk_remote)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/build/reproducible-path/python-mitogen-0.3.45/mitogen/parent.py", line 2029, in call
    return self.default_call_chain.call(fn, *args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/build/reproducible-path/python-mitogen-0.3.45/mitogen/parent.py", line 1986, in call
    return receiver.get().unpickle(throw_dead=False)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/build/reproducible-path/python-mitogen-0.3.45/mitogen/core.py", line 986, in unpickle
    return next(self.unpickle_iter(throw, throw_dead, find_class))
  File "/build/reproducible-path/python-mitogen-0.3.45/mitogen/core.py", line 1027, in unpickle_iter
    raise obj
mitogen.core.CallError: builtins.AssertionError: 
  File "<stdin>", line 3975, in _dispatch_one
  File "master:/build/reproducible-path/python-mitogen-0.3.45/tests/router_test.py", line 424, in test_s
iblings_cant_talk_remote
    test_siblings_cant_talk(econtext.router)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "master:/build/reproducible-path/python-mitogen-0.3.45/tests/router_test.py", line 418, in test_siblings_cant_talk
    assert 'routing mode prevents forward of ' in logs.stop()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Full build log: https://debusine.debian.net/debian/developers/artifact/3522629/

### Mitogen version

0.3.45

### Ansible version (if applicable)

_No response_

### OS and environment

Debian unstable amd64

### Steps to reproduce

Steps:
1. `python3.14 -m unittest discover -v --start-directory tests --pattern '*_test.py'`

### Anything else

_No response_

With `MITOGEN_LOG_LEVEL=debug`
```console
➜  mitogen2 git:(master) MITOGEN_LOG_LEVEL=debug .venv/bin/python3.14 -m unittest discover --start-directory tests --pattern router_test.py
............................
----------------------------------------------------------------------
Ran 28 tests in 4.444s

OK
```

Without
```console
➜  mitogen2 git:(master) .venv/bin/python3.14 -m unittest discover --start-directory tests --pattern router_test.py
.....FF.........F.........FE
======================================================================
ERROR: test_siblings_cant_talk_parent (router_test.UnidirectionalTest.test_siblings_cant_talk_parent)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 433, in test_siblings_cant_talk_parent
    parent.call(test_siblings_cant_talk_remote)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alex/src/mitogen2/mitogen/parent.py", line 2029, in call
    return self.default_call_chain.call(fn, *args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
  File "/Users/alex/src/mitogen2/mitogen/parent.py", line 1986, in call
    return receiver.get().unpickle(throw_dead=False)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
  File "/Users/alex/src/mitogen2/mitogen/core.py", line 986, in unpickle
    return next(self.unpickle_iter(throw, throw_dead, find_class))
  File "/Users/alex/src/mitogen2/mitogen/core.py", line 1027, in unpickle_iter
    raise obj
mitogen.core.CallError: builtins.AssertionError:
  File "<stdin>", line 3975, in _dispatch_one
  File "master:/Users/alex/src/mitogen2/tests/router_test.py", line 420, in test_siblings_cant_talk_remote
    test_siblings_cant_talk(econtext.router)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "master:/Users/alex/src/mitogen2/tests/router_test.py", line 414, in test_siblings_cant_talk
    assert 'routing mode prevents forward of ' in logs.stop()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


======================================================================
FAIL: test_local_dead_message (router_test.MessageSizeTest.test_local_dead_message)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 284, in test_local_dead_message
    self.assertIn(expect, logs.stop())
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'message too large (max 4096 bytes)' not found in "Router(Broker(3380)): registering Context(0, 'master') to stream <Stream parent #efd0>\nPython version is 3.14.3 (main, Mar  3 2026, 15:46:55) [Clang 21.1.4 ]\nParent is context 0 (master); my ID is 1\npid:72882 ppid:72877 uid:501/501, gid:20/20 host:'kintha'\nRecovered sys.executable: '/Users/alex/src/mitogen2/.venv/bin/python3.14'\n"

======================================================================
FAIL: test_local_exceeded (router_test.MessageSizeTest.test_local_exceeded)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 266, in test_local_exceeded
    self.assertIn(expect, logs.stop())
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'message too large (max 4096 bytes)' not found in ''

======================================================================
FAIL: test_refuse_all (router_test.PolicyTest.test_refuse_all)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 167, in test_refuse_all
    self.assertIn(self.router.refused_msg, log.stop())
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 'refused by policy' not found in ''

======================================================================
FAIL: test_siblings_cant_talk_master (router_test.UnidirectionalTest.test_siblings_cant_talk_master)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 426, in test_siblings_cant_talk_master
    test_siblings_cant_talk(self.router)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/Users/alex/src/mitogen2/tests/router_test.py", line 414, in test_siblings_cant_talk
    assert 'routing mode prevents forward of ' in logs.stop()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

----------------------------------------------------------------------
Ran 28 tests in 4.312s

FAILED (failures=4, errors=1)
```

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -23,6 +23,7 @@ In progress (unreleased)
 
 * :gh:issue:`1502` :mod:`mitogen`: Preserve :py:mod:`logging` config in forked
   children
+* :gh:issue:`1499` tests: Fix uses of ``testlib.LogCapturer``
 
 
 v0.3.45 (2026-03-30)

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -24,6 +24,8 @@ In progress (unreleased)
 * :gh:issue:`1502` :mod:`mitogen`: Preserve :py:mod:`logging` config in forked
   children
 * :gh:issue:`1499` tests: Fix uses of ``testlib.LogCapturer``
+* :gh:issue:`1499` tests: Run Mitogen unittests without
+  ``MITOGEN_LOG_LEVEL=debug``
 
 
 v0.3.45 (2026-03-30)

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -21,6 +21,9 @@ To avail of fixes in an unreleased version, please download a ZIP file
 In progress (unreleased)
 ------------------------
 
+* :gh:issue:`1502` :mod:`mitogen`: Preserve :py:mod:`logging` config in forked
+  children
+
 
 v0.3.45 (2026-03-30)
 --------------------

--- a/mitogen/fork.py
+++ b/mitogen/fork.py
@@ -51,6 +51,8 @@ LOG = logging.getLogger(__name__)
 # fork is brainwrong from the stone age.
 FORK_SUPPORTED = sys.version_info >= (2, 6)
 
+HAVE_OS_REGISTER_AT_FORK = hasattr(os, 'register_at_fork')
+
 
 class Error(mitogen.core.StreamError):
     pass
@@ -88,8 +90,9 @@ def on_fork():
     Should be called by any program integrating Mitogen each time the process
     is forked, in the context of the new child.
     """
-    reset_logging_framework()  # Must be first!
-    fixup_prngs()
+    if not HAVE_OS_REGISTER_AT_FORK:
+        reset_logging_framework()  # Must be first!
+        fixup_prngs()
     mitogen.core.Latch._on_fork()
     mitogen.core.Side._on_fork()
     mitogen.core.ExternalContext.service_stub_lock = threading.Lock()

--- a/tests/create_child_test.py
+++ b/tests/create_child_test.py
@@ -85,7 +85,7 @@ def wait_read(fp, n):
         poller.start_receive(fp.fileno())
         for _ in poller.poll():
             return os.read(fp.fileno(), n)
-        assert False
+        raise EnvironmentError('Nothing was ready to read on %r' % (fp,))
     finally:
         poller.close()
 

--- a/tests/fork_test.py
+++ b/tests/fork_test.py
@@ -1,6 +1,5 @@
 import os
 import random
-import subprocess
 import sys
 import unittest
 
@@ -24,6 +23,7 @@ import mitogen.core
 import mitogen.fork
 
 import testlib
+from testlib import subprocess
 
 
 def _find_ssl_linux():
@@ -31,28 +31,32 @@ def _find_ssl_linux():
     if ssl_object_path is None:
         # No __file__ because it's builtin
         ssl_object_path = sys.executable
-    proc = subprocess.Popen(
+    proc = subprocess.run(
         ['ldd', ssl_object_path],
         stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+        check=True,
     )
-    b_stdout, b_stderr = proc.communicate()
-    assert proc.returncode == 0
-    assert b_stderr.decode() == ''
-    for line in b_stdout.decode().splitlines():
+    if proc.stderr:
+        raise subprocess.CalledProcessError(
+            proc.returncode, proc.args, proc.stdout, proc.stderr,
+        )
+    for line in proc.stdout.decode().splitlines():
         bits = line.split()
         if bits[0].startswith('libssl'):
             return bits[2]
 
 
 def _find_ssl_darwin():
-    proc = subprocess.Popen(
+    proc = subprocess.run(
         ['otool', '-l', _ssl.__file__],
         stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+        check=True,
     )
-    b_stdout, b_stderr = proc.communicate()
-    assert proc.returncode == 0
-    assert b_stderr.decode() == ''
-    for line in b_stdout.decode().splitlines():
+    if proc.stderr:
+        raise subprocess.CalledProcessError(
+            proc.returncode, proc.args, proc.stdout, proc.stderr,
+        )
+    for line in proc.stdout.decode().splitlines():
         bits = line.split()
         if bits[0] == 'name' and 'libssl' in bits[1]:
             return bits[1]
@@ -81,7 +85,8 @@ def random_random():
 
 def RAND_pseudo_bytes(n=32):
     buf = ctypes.create_string_buffer(n)
-    assert 1 == c_ssl.RAND_pseudo_bytes(buf, n)
+    if c_ssl.RAND_pseudo_bytes(buf, n) != 1:
+        raise ValueError
     return buf[:]
 
 

--- a/tests/log_handler_test.py
+++ b/tests/log_handler_test.py
@@ -123,11 +123,11 @@ class LogRecordFactoryTest(testlib.RouterMixin, testlib.TestCase):
             return record
 
         logging_setLogRecordFactory(record_factory)
-        c1 = self.router.local(name="c1")
         log = testlib.LogCapturer(
-            __name__, formatter=logging.Formatter("%(custom_attribute)x - %(message)s")
+            formatter=logging.Formatter("%(custom_attribute)x - %(message)s"),
         )
         log.start()
+        c1 = self.router.local(name="c1")
         c1.call(log_test)
         logs = log.stop()
         self.assertIn("deadbeef - This is a test", logs)

--- a/tests/minify_test.py
+++ b/tests/minify_test.py
@@ -79,17 +79,18 @@ class MitogenCoreTest(testlib.TestCase):
         mlines = minified.splitlines()
         for i, (orig, mini) in enumerate(zip(olines, mlines)):
             if i < 2:
-                assert orig == mini
+                self.assertEqual(orig, mini)
                 continue
 
             owords = orig.split()
             mwords = mini.split()
-            assert len(mwords) == 0 or (mwords[0] == owords[0]), pprint.pformat({
-                'line': i+1,
-                'filename': name,
-                'owords': owords,
-                'mwords': mwords,
-            })
+            self.assertTrue(
+                len(mwords) == 0 or (mwords[0] == owords[0]),
+                pprint.pformat({
+                    'line': i+1, 'name': name,
+                    'owords': owords, 'mwords': mwords,
+                }),
+            )
 
     PY_24_25_SKIP = [
         # cProfile unsupported on 2.4, 2.6+ syntax is fine here.

--- a/tests/responder_test.py
+++ b/tests/responder_test.py
@@ -179,9 +179,9 @@ class BrokenModulesTest(testlib.TestCase):
 
 class ForwardTest(testlib.RouterMixin, testlib.TestCase):
     def test_forward_to_nonexistent_context(self):
-        nonexistent = mitogen.core.Context(self.router, 123)
         capture = testlib.LogCapturer()
         capture.start()
+        nonexistent = mitogen.core.Context(self.router, 123)
         self.broker.defer_sync(lambda:
             self.router.responder.forward_modules(
                 nonexistent,

--- a/tests/router_test.py
+++ b/tests/router_test.py
@@ -253,10 +253,9 @@ class MessageSizeTest(testlib.BrokerMixin, testlib.TestCase):
     klass = mitogen.master.Router
 
     def test_local_exceeded(self):
-        router = self.klass(broker=self.broker, max_message_size=4096)
-
         logs = testlib.LogCapturer()
         logs.start()
+        router = self.klass(broker=self.broker, max_message_size=4096)
 
         # Send message and block for one IO loop, so _async_route can run.
         router.route(mitogen.core.Message.pickled(' '*8192))
@@ -266,12 +265,10 @@ class MessageSizeTest(testlib.BrokerMixin, testlib.TestCase):
         self.assertIn(expect, logs.stop())
 
     def test_local_dead_message(self):
-        # Local router should generate dead message when reply_to is set.
-        router = self.klass(broker=self.broker, max_message_size=4096)
-
         logs = testlib.LogCapturer()
         logs.start()
-
+        # Local router should generate dead message when reply_to is set.
+        router = self.klass(broker=self.broker, max_message_size=4096)
         expect = router.too_large_msg % (4096,)
 
         # Try function call. Receiver should be woken by a dead message sent by
@@ -397,9 +394,6 @@ class NoRouteTest(testlib.RouterMixin, testlib.TestCase):
 def test_siblings_cant_talk(router):
     l1 = router.local()
     l2 = router.local()
-    logs = testlib.LogCapturer()
-    logs.start()
-
     try:
         l2.call(ping_context, l1)
     except mitogen.core.CallError:
@@ -410,27 +404,36 @@ def test_siblings_cant_talk(router):
         l1.context_id,
         mitogen.context_id,
     )
-    assert msg in str(e)
-    assert 'routing mode prevents forward of ' in logs.stop()
+    return (msg, e)
 
 
 @mitogen.core.takes_econtext
 def test_siblings_cant_talk_remote(econtext):
     mitogen.parent.upgrade_router(econtext)
-    test_siblings_cant_talk(econtext.router)
+    return test_siblings_cant_talk(econtext.router)
 
 
 class UnidirectionalTest(testlib.RouterMixin, testlib.TestCase):
     def test_siblings_cant_talk_master(self):
+        capture = testlib.LogCapturer()
+        capture.start()
         self.router.unidirectional = True
-        test_siblings_cant_talk(self.router)
+        expected_msg, exc = test_siblings_cant_talk(self.router)
+        logs = capture.stop()
+        self.assertIn(expected_msg, str(exc))
+        self.assertIn('routing mode prevents forward of ', logs)
 
     def test_siblings_cant_talk_parent(self):
         # ensure 'unidirectional' attribute is respected for contexts started
         # by children.
+        capture = testlib.LogCapturer()
+        capture.start()
         self.router.unidirectional = True
         parent = self.router.local()
-        parent.call(test_siblings_cant_talk_remote)
+        expected_msg, exc = parent.call(test_siblings_cant_talk_remote)
+        logs = capture.stop()
+        self.assertIn(expected_msg, str(exc))
+        self.assertIn('routing mode prevents forward of ', logs)
 
     def test_auth_id_can_talk(self):
         self.router.unidirectional = True

--- a/tests/testlib.py
+++ b/tests/testlib.py
@@ -58,7 +58,7 @@ TESTS_DIR =                     os.path.join(os.path.dirname(__file__))
 DATA_DIR =                      os.path.join(TESTS_DIR, 'data')
 TESTMODS_DIR =                  os.path.join(TESTS_DIR, 'testmods')
 
-if mitogen.is_master:
+if mitogen.is_master and 'MITOGEN_LOG_LEVEL' in os.environ:
     mitogen.utils.log_to_file()
 
 if faulthandler is not None:

--- a/tests/testlib.py
+++ b/tests/testlib.py
@@ -316,21 +316,11 @@ def log_fd_calls():
     os.dup = dup
 
 
-class CaptureStreamHandler(logging.StreamHandler):
-    def __init__(self, *args, **kwargs):
-        logging.StreamHandler.__init__(self, *args, **kwargs)
-        self.msgs = []
-
-    def emit(self, msg):
-        self.msgs.append(msg)
-        logging.StreamHandler.emit(self, msg)
-
-
 class LogCapturer(object):
     def __init__(self, name=None, formatter=None):
         self.sio = StringIO()
         self.logger = logging.getLogger(name)
-        handler = CaptureStreamHandler(self.sio)
+        handler = logging.StreamHandler(self.sio)
         if formatter is not None:
             handler.setFormatter(formatter)
         self.handler = handler
@@ -349,9 +339,6 @@ class LogCapturer(object):
         if isinstance(s, mitogen.core.BytesType):
             s = s.decode('utf-8')
         return s
-
-    def msgs(self):
-        return self.handler.msgs
 
     def __enter__(self):
         self.start()

--- a/tests/testlib.py
+++ b/tests/testlib.py
@@ -324,14 +324,11 @@ class LogCapturer(object):
         if formatter is not None:
             handler.setFormatter(formatter)
         self.handler = handler
-        self.old_propagate = self.logger.propagate
-        self.old_handlers = self.logger.handlers
         self.old_level = self.logger.level
 
     def start(self):
-        self.logger.handlers = [self.handler]
-        self.logger.propagate = False
-        self.logger.level = logging.DEBUG
+        self.logger.addHandler(self.handler)
+        self.logger.setLevel(logging.DEBUG)
 
     def raw(self):
         s = self.sio.getvalue()
@@ -348,9 +345,8 @@ class LogCapturer(object):
         self.stop()
 
     def stop(self):
-        self.logger.level = self.old_level
-        self.logger.handlers = self.old_handlers
-        self.logger.propagate = self.old_propagate
+        self.logger.setLevel(self.old_level)
+        self.logger.removeHandler(self.handler)
         return self.raw()
 
 

--- a/tests/timer_test.py
+++ b/tests/timer_test.py
@@ -157,7 +157,7 @@ class CancelTest(TimerListMixin, testlib.TestCase):
 
 @mitogen.core.takes_econtext
 def do_timer_test_econtext(econtext):
-    do_timer_test(econtext.broker)
+    return do_timer_test(econtext.broker)
 
 
 def do_timer_test(broker):
@@ -169,9 +169,7 @@ def do_timer_test(broker):
             lambda: latch.put('hi'),
         )
     )
-
-    assert 'hi' == latch.get()
-    assert mitogen.core.now() > (now + 0.250)
+    return (latch.get(), now, mitogen.core.now())
 
 
 class BrokerTimerTest(testlib.TestCase):
@@ -180,17 +178,21 @@ class BrokerTimerTest(testlib.TestCase):
     def test_call_later(self):
         broker = self.klass()
         try:
-            do_timer_test(broker)
+            response, t0, t1 = do_timer_test(broker)
         finally:
             broker.shutdown()
             broker.join()
+        self.assertEqual(response, 'hi')
+        self.assertTrue(t1-t0 >= 0.25)
 
     def test_child_upgrade(self):
         router = mitogen.master.Router()
         try:
             c = router.local()
             c.call(mitogen.parent.upgrade_router)
-            c.call(do_timer_test_econtext)
+            response, t0, t1 = c.call(do_timer_test_econtext)
         finally:
             router.broker.shutdown()
             router.broker.join()
+        self.assertEqual(response, 'hi')
+        self.assertTrue(t1-t0 >= 0.25)

--- a/tox.ini
+++ b/tox.ini
@@ -140,7 +140,6 @@ setenv =
     ans13: MITOGEN_TEST_DISTRO_SPECS=alma9-py3 debian12-py3 ubuntu2404-py3
     m_ans: ANSIBLE_SKIP_TAGS=resource_intensive
     m_lcl: ANSIBLE_SKIP_TAGS=issue_776,resource_intensive
-    m_mtg: MITOGEN_LOG_LEVEL=debug
     s_lin: ANSIBLE_STRATEGY=linear
 allowlist_externals =
     # Added: Tox 3.18: Tox 4.0+

